src/backend/distributed/master/master_repair_shards.c

```diff
@@ -488,3 +550,44 @@ WorkerApplyShardDDLCommandList(List *ddlCommandList, int64 shardId)
 
 	return applyDdlCommandList;
 }
+
+
+/*
+ * LockReferencedReferenceShardDistributionMetadata takes the given lock
+ * on the reference tables which has a foreign key from the given relation.
+ */
+static void
+LockReferencedReferenceShardDistributionMetadata(Oid relationId, int lock)
+{
+	List *referencedTableList = NIL;
+	ListCell *referencedTableCell = NULL;
+
+	referencedTableList = GetReferencedReferenceTableList(relationId);
+	referencedTableList = SortList(referencedTableList, CompareOids);
+
+	foreach(referencedTableCell, referencedTableList)
+	{
+		Oid referencedTableId = lfirst_oid(referencedTableCell);
+		List *shardList = NIL;
+		uint64 *shardIdPointer = NULL;
+		Oid referencedShardId = InvalidOid;
+
+		if (PartitionMethod(referencedTableId) != DISTRIBUTE_BY_NONE)
+		{
+			continue;
+		}
+
+		/* check that user has owner rights in referenced table */
+		EnsureTableOwner(referencedTableId);
+
+		/*
+		 * Block concurrent reads on the referenced table so that there
+		 * is no inconsistency between before/after copy.
+		 */
+		shardList = LoadShardList(referencedTableId);
+		shardIdPointer = (uint64 *) linitial(shardList);
+		referencedShardId = (*shardIdPointer);
+
+		LockShardDistributionMetadata(referencedShardId, lock);
```

Comment:
even if it's not yet very useful, it'd be nice to also do `LockShardListMetadataOnWorkers` here to (in the future) block writes from nodes with metadata